crowdsourcehinter/static/js/src/crowdsourcehinter.js

```diff
@@ -0,0 +1,305 @@
+     * is triggered by clicking the "contribute a new hint" button.
+     * @param clicked is the "contribute a new hint" button that was clicked
+     */
+    function create_text_input(){ return function(clicked){
+        $('.csh_student_hint_creation', element).each(function(){
+            $(clicked.currentTarget).show();
+        });
+        $('.csh_student_text_input', element).remove();
+        $('.csh_submit_new', element).remove();
+        $(clicked.currentTarget).hide();
+        student_answer = $(clicked.currentTarget).parent().parent().find('.csh_answer_text').attr('answer');
+        $(".csh_student_answer", element).each(function(){
+            if ($('.csh_answer_text', element).attr('answer') == student_answer){
+                var html = "";
+                $(function(){
+                    var template = $('#student_hint_creation').html();
+                    var data = {student_answer: student_answer};
+                    html = Mustache.render(template, data); 
+                });
+                $(this).append(html);
+            }
+        });
+    }}
+    $(element).on('click', '.csh_student_hint_creation', create_text_input($(this)));
+
+    /**
+     * Submit a new hint created by the student to the hint pool. Hint text is in
+     * the text input area created by create_text_input. Contributed hints are specific to 
+     * incorrect answers. Triggered by clicking the "submit hint" button.
+     * @param clicked is the "submit hint" button clicked
+     */
+    function submit_new_hint(){ return function(clicked){
+    //add the newly created hint to the hinter's pool of hints
+        if($('.csh_student_text_input', element).val().length > 0){
+            var answerdata = unescape(clicked.currentTarget.attributes['answer'].value);
+            var newhint = unescape($('.csh_student_text_input').val());
+            $('.csh_submitbutton', element).show();
+            $.ajax({
+                type: "POST",
+                url: runtime.handlerUrl(element, 'add_new_hint'),
+                data: JSON.stringify({"submission": newhint, "answer": answerdata}),
+                success: Logger.log('crowd_hinter.submit_new.click.event', {"student_answer": answerdata, "new_hint_submission": newhint})
+            });
+            $('.csh_student_text_input', element).remove();
+            $(clicked.currentTarget).remove();
+        }
+    }}
+    $(element).on('click', '.csh_submit_new', submit_new_hint($(this)));
+
+    /**
+     * Send vote data to modify a hint's rating (or mark it as reported). Triggered by
+     * clicking a button to upvote, downvote, or report the hint (both before and after
+     * the student correctly submits an answer).
+     * @param clicked is the rate_hint button clicked (upvote/downvote/report)
+     */
+    function rate_hint(){ return function(clicked){
+        rating = clicked.currentTarget.attributes['data-rate'].value;
+        if(!voted || rating=="report"){
```

Comment:
1. voted is the wrong variable name if it includes flagged
2. It's helpful to document why you have this check/what it does.